lib/spack/spack/package.py

```diff
@@ -1129,6 +1133,34 @@ def do_patch(self):
         else:
             touch(no_patches_file)
 
+    def patches_to_apply(self):
+        """If the patch set does not change between two invocations of spack,
+           then all patches in the set will be applied in the same order"""
+        patchesToApply = itertools.chain.from_iterable(
+            patch_list
+            for spec, patch_list in self.patches.iteritems()
+            if self.spec.satisfies(spec))
+        return sorted(patchesToApply, key=lambda p: p.path_or_url)
+
+    def content_hash(self, remove_pkg_name=False):
+        """Create a hash based on the sources and logic used to build the
+        package. This includes the contents of all applied patches and the
+        contents of applicable functions in the package subclass."""
+        hashContent = list()
+        # TODO? in cases where a digest or source_id isn't available, should
+        # this attempt to download the source and set one? For URL resources
+        # this will involve downloading the entire resource (repositories
+        # typically allow for just retrieving the latest commit ID)
+        source_id = fs.for_package_version(self, self.version).source_id()
+        if not source_id:
+            return
+        hashContent.append(source_id)
+        hashContent.extend(':'.join((p.file_hash, str(p.level)))
+                           for p in self.patches_to_apply())
+        hashContent.append(package_hash(self.spec, remove_pkg_name))
+        return base64.b32encode(
+            hashlib.sha1(''.join(sorted(hashContent))).digest()).lower()
```

Comment:
Default this to `sha256`.  I realize we'll run it through a `sha1` eventually right now, but I want the default choice to be `sha256` going forward.